replace password with unix socket authentication for root user

DanScharon opened the pull request:
this pull request implements two things:
- load differences in installation through distribution dependent variables
- replace password authentication for the database root user with authentication via unix socket

authentication via unix socket has been set as the default for the MariaDB user  'root'@'localhost' since MariaDB 10.4.3
Since then there is no need anymore to set a database root user password and it is actually better to not do so because it can have bad side affects, e.g. in Debian/Ubuntu some maintainer scripts rely on this function and also `automysqlbackup`. For the latter you would have to store again a password in a cleartext file.
For  the creation of the opencast user and database it is sufficient to use `become` for root permissions.

`database_root_password` is still left in as optional because some users may provision to an already installed MariaDB which has a database root user password already set.

DanScharon committed:
make database_root_password optional and use unix_socket as default auth plugin, ensure that the tasks run for both Debian and RedHat based distros

DanScharon committed:
make database_root_password optional and use unix_socket as default auth plugin, ensure that the tasks run for both Debian and RedHat based distros

DanScharon committed:
provide a default value for database_root_password and update README.md accordingly

DanScharon committed:
if auth plugin is currently not unix_socket we need database_root_password ;)

lkiesow committed:
MariaDB Socket Configuration for el9

This role fails on el9 systems.
Setting a socket location fixes the problem.

lkiesow committed:
Run Molecule Tests on LXC

This patch fixes the Molecule tests by running the playbooks on LXC
containers in GitHub Actions.

lkiesow committed:
Mark as ready for el9

This patch runs the tests also on el9 and marks the role as being
compatible with that version.

lkiesow commented on defaults/main.yml:
```diff
@@ -5,6 +5,8 @@ database_name: opencast
 database_user_name: opencast
 # mariadb root user
 database_root_user: root
+# mariadb root user password
+database_root_password: '4567'
```
There should be an explanation that this is now only being used for authentication and that this password is not set

DanScharon replied:
I had already adapted the README but I have added additional context in `defaults/main.yml` file now

lkiesow commented on tasks/main.yml:
```diff
@@ -1,19 +1,18 @@
-      - mariadb-client
-      - python3-pymysql
+- name: Loading platform variables
+  include_vars: "{{ item }}"
+  with_first_found:
+    - "{{ ansible_distribution }}-{{ ansible_distribution_release }}.yml"
+    - "{{ ansible_distribution }}.yml"
+    - "{{ ansible_os_family }}.yml"
+  tags: always
```
Why this complicated? We know that only the last will ever catch, won't it?

```suggestion
  include_vars: "{{ ansible_os_family }}.yml"
  tags: always
```

DanScharon replied:
agreed. I had put the others there as a preparation for possible further distribution deviations, but there seem to be no more

lkiesow commented on tasks/main.yml:
```diff
@@ -1,19 +1,18 @@
+  tags: always
+
+- name: Install MariaDB
+  ansible.builtin.package:
+    name: "{{ item }}"
+    state: present
+  with_items: "{{ mariadb_packages }}"
```
Shouldn't this also work (and doesn't require a loop):

```suggestion
  ansible.builtin.package:
    name: "{{ mariadb_packages }}"
    state: present
```

As an alternative, you could even use a simple inline switch here.
That way, you wouldn't have to look up somewhere else what will be installed:

```yaml
- name: Install MariaDB on RedHat Systems
  ansible.builtin.package:
    name:
      - mariadb-server
      - mariadb
      - python3-{{ 'PyMySQL' if ansible_os_family == 'RedHat' else 'pymysql' }}
```

DanScharon replied:
Yes, this also works. Not every module accepts lists as parameters so I had used a loop here, but the package module apparently does :) for inline switches I would have to use several because not only the python package name differs and then I can just put it into a variable because we need a distribution distinction anyways

lkiesow commented on tasks/main.yml:
```diff
@@ -39,7 +38,7 @@
   notify: Restart MariaDB
   ansible.builtin.ini_file:
-    dest: /etc/my.cnf.d/opencast.cnf
+    dest: "{{ mariadb_configuration_file }}"
```
Multiple places to comment on this, but let me make the comment here once.

Why set this to different places on Debian and Red Hat? The previous location should be fine. Setting this to something else will just cause confusion for people upgrading from an older version of this role and I don't see any benefit.

DanScharon replied:
Because `/etc/my.cnf.d` as a config inclusion path does not work on Debian and derivatives. The path does not exist by default and mariadb in Debian does not look for it. So if people upgrade from an older role version it hadn't worked anyways before in Debian and derivatives.

lkiesow commented on tasks/main.yml:
```diff
@@ -98,30 +80,34 @@
+    login_unix_socket: /run/mysqld/mysqld.sock
 
 - name: Ensure root user can only connect from the local machine
+  community.mysql.mysql_query:
```
Does this matter at all if the next step is to force the `root` user to use Unix sockets?

DanScharon replied:
Yes, because even after you have forced the `root` user to use the unix_socket plugin you'd also want to clean up entries that still refer to remote hosts because after a (accidental) deactivation of the plugin they'd become active/relevant again

lkiesow reviewed (changes requested):
Sorry, we seem to have overlooked this pull request.
I've left a review and hope we can get it in soon.

Can you rebase this onto the current `main` branch, make sure that the tests pass and take a look at the comments?

lkiesow committed:
Fix os version specification

This patch fixes the specification for supported `EL` versions which had
a typo in the dictionary key.

DanScharon committed:
use a separate file for config adjustments and not the standard server conf file so that it survives package upgrades

DanScharon committed:
simplify tasks and add contextual explanation

DanScharon commented:
> Sorry, we seem to have overlooked this pull request. I've left a review and hope we can get it in soon.
> 
> Can you rebase this onto the current `main` branch, make sure that the tests pass and take a look at the comments?

Thank you for the review! I have incorporated your feedback now

tibroc commented on tasks/main.yml:
```diff
@@ -90,6 +65,8 @@
     login_user: "{{ database_root_user }}"
     login_password: "{{ database_root_password }}"
+    check_implicit_admin: true
+    login_unix_socket: /run/mysqld/mysqld.sock
```
This needs a os-specific variable as well. E.g. on CentOS Stream 9 the defautl socket location is `/var/lib/mysql/mysql.sock`.

tibroc commented on tasks/main.yml:
```diff
@@ -98,30 +75,32 @@
   community.mysql.mysql_user:
     user: ""
-    host: "{{ item }}"
+    host_all: yes
```
```suggestion
    host_all: true
```

tibroc reviewed (changes requested):
I tested this on two different oparating systems (Debian 12 and CentOS Stream 9) and it seems to work fine, *if* you adjust the socket location accordingly (see comment). I think this is a good improvement of the role and we should merge this (and sorry again for the delay). Thank you @DanScharon !

DanScharon committed:
make socket path os dependent as well, linting

DanScharon committed:
Merge remote-tracking branch 'origin/main' into unix_socket

# Conflicts:
#	.dev_requirements.txt
#	.github/workflows/molecule.yml
#	meta/main.yml
#	molecule/default/Dockerfile.j2
#	molecule/default/molecule.yml
#	molecule/default/verify.yml
#	tasks/main.yml

DanScharon commented:
closing in favor of https://github.com/elan-ev/opencast_mariadb/pull/9
